### DISCUSSION TITLE
How can you determine the parent asset which triggered a materialization of a child asset?

### DISCUSSION THREAD

**Discussion by dhirschfeld** [dagster-io/dagster#20530] [Q&A] [answered]
I have two assets `A` and `B` and a third asset `C` which depends on both `A` & `B`.

`C` has an auto-materialization policy which will be triggered on updates to *either* parent e.g. 

```python
@asset
def asset_b(context):
	pass


@asset
def asset_b(context):
	pass


@asset(
    deps=[
        asset_a,
		asset_b,
    ],
    auto_materialize_policy=(
        AutoMaterializePolicy
        .eager(None)
        .without_rules(
            AutoMaterializeRule.skip_on_parent_missing(),
        )
    ),
)
def asset_c(context):
	# determine which dependcency, asset_a or asset_b, has triggered this materialization
	pass
```

From within `asset_c` how can I tell which dependency update triggered the auto-materialization policy to materialize `asset_c`

**Comment by dhirschfeld** [comment]
Here's my best attempt:
```python
def get_run_ids_for_dependencies(
    ctx: AssetExecutionContext,
) -> dict[str, str | None]:
    """Returns the `dagster` `run_ids` for the dependencies of the specified
    `ctx`.

    Parameters
    ----------
    ctx : AssetExecutionContext
        The `dagster` context object for the asset materialization.

    Returns
    -------
    run_ids : dict[str, str | None]:
        A mapping of dependency name to `run_id` for all dependencies of the
        specified `ctx`. If a dependency didn't trigger the asset
        materialization, it's `run_id` will be `None`.

    """
    run_ids = {}
    for asset_key in ctx.assets_def.asset_deps[ctx.asset_key]:
        key = '/'.join(asset_key.path)
        provenance = ctx.get_asset_provenance(asset_key)
        run_ids[key] = provenance.code_version if provenance else None
    return run_ids
```

Is this correct?
Is there no easier way?

**Comment by sryza** [comment, marked as answer]
There currently isn't a straightforward way to determine this. If the approach you listed works for you, I'd recommend going with it.  Another option could be to find the latest materializations of the upstream assets and look at which one occurred the latest.

**Reply by dhirschfeld** [reply]
I *think* that works in that one or more dependencies will have an associated `run_id` if they have been updated since the the previous materialization of the child asset.

What I'll do in `asset_c` is to loop over the `run_ids` for the parent assets (dependencies) and process the new data from each.

**Reply by ogerardin** [reply]
> Another option could be to find the latest materializations of the upstream assets and look at which one occurred the latest.

Here's code that implements this option:

```
def latest_materialization_timestamp(dagster_instance: DagsterInstance,
                                     asset_key: AssetKey):
    event = dagster_instance.get_latest_materialization_event(asset_key)
    return event.timestamp if event is not None else None


def most_recently_materialized_asset(dagster_instance: DagsterInstance,
                                     asset_keys: Iterable[AssetKey]) -> AssetKey:
    asset_timestamps = {
        asset_key: dagster_instance.get_latest_materialization_event(asset_key).timestamp for asset_key in asset_keys
    }
    return max(asset_timestamps, key=asset_timestamps.get)

```

In your asset function you can call

`asset_key = most_recently_materialized_asset(context.instance, [ASSET_KEY_1, ASSET_KEY_2, ...])`

**Comment by dhirschfeld** [comment]
Which of your dependencies triggered the materialization (i.e. has new data) seems like information you'd want to know so it might be worth a helper function on the `context`?
